pkg/yurtappmanager/controller/nodepool/nodepool_controller.go

```diff
@@ -0,0 +1,600 @@
+/*
+Copyright 2020 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodepool
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"reflect"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	nodeutil "k8s.io/kubernetes/pkg/controller/util/node"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	appsv1alpha1 "github.com/alibaba/openyurt/pkg/yurtappmanager/apis/apps/v1alpha1"
+	"github.com/alibaba/openyurt/pkg/yurtappmanager/constant"
+	"github.com/alibaba/openyurt/pkg/yurtappmanager/util/gate"
+)
+
+const controllerName = "nodepool-controller"
+
+var concurrentReconciles = 3
+
+// NodePoolReconciler reconciles a NodePool object
+type NodePoolReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	recorder          record.EventRecorder
+	createDefaultPool bool
+}
+
+type NodePoolRelatedAttributes struct {
+	Labels      map[string]string `json:"labels,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+	Taints      []corev1.Taint    `json:"taints,omitempty"`
+}
+
+// Add creates a new NodePool Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, ctx context.Context) error {
```

Comment:
Who calls this method？

Reply:
append `nodepool.Add` to the `ControllerFuncs`.